world/physics/RigidBody.cpp

```diff
@@ -0,0 +1,206 @@
+                     const Quaternion& init_g_q_b, const Vector3& init_vel_b_wrt_g_in_b,
+                     const Vector3& init_omega_b_wrt_g_in_b)
+    : mass_(mass), InertiaTensor_(InertiaTensor) {
+    InertiaTensorInverse_ = InertiaTensor.inverse();
+
+    set_pos_b_wrt_g_in_g(init_pos_b_wrt_g_in_g);
+    set_g_q_b(init_g_q_b);
+    set_vel_b_wrt_g_in_b(init_vel_b_wrt_g_in_b);
+    set_omega_b_wrt_g_in_b(init_omega_b_wrt_g_in_b);
+
+    u_ = Vector6::Zero();
+}
+
+Vector3 RigidBody::get_gravity_b() {
+    // TODO(Amaar): gravity and J2 computations here
+    return Vector3::Zero();
+}
+
+void RigidBody::applyForce(const Vector3& force_b, const Vector3& pointOfApplication_b) {
+    set_net_force_b(get_net_force_b() + force_b);
+
+    // accounting for the applied body force's resulting moment
+    applyMoment(pointOfApplication_b.cross(force_b));
+}
+
+void RigidBody::applyMoment(const Vector3& moment_b) {
+    set_net_moment_b(get_net_moment_b() + moment_b);
+}
+
+void RigidBody::clearAppliedForcesAndMoments() {
+    set_net_force_b(Vector3::Zero());
+    set_net_moment_b(Vector3::Zero());
+}
+
+StateVector f(const StateVector& x, const Vector6& u, const Matrix_3x3& InertiaTensor,
+              const Matrix_3x3& InertiaTensorInverse, double mass) {
+    StateVector xdot = StateVector::Zero();   // derivative of the state vector
+
+    // aliases for brevity
+    auto r           = ::get_pos_b_wrt_g_in_g(x);
+    auto v_b         = ::get_vel_b_wrt_g_in_b(x);
+    auto q           = ::get_g_q_b(x);
+    auto f_b         = ::get_net_force_b(u);
+    const auto& J    = InertiaTensor;
+    const auto& Jinv = InertiaTensorInverse;
+    auto tau_b       = ::get_net_moment_b(u);
+    auto w           = ::get_omega_b_wrt_g_in_b(x);
+
+    Matrix_4x4 L = Matrix_4x4::Zero();
+    L << q.w(), -q.x(), -q.y(), -q.z(),   //
+        q.x(), q.w(), -q.z(), q.y(),      //
+        q.y(), q.z(), q.w(), -q.x(),      //
+        q.z(), -q.y(), q.x(), q.w();
+
+    Matrix_4x3 H = Matrix_4x3::Zero();
+    H(1, 0)      = 1;
+    H(2, 1)      = 1;
+    H(3, 2)      = 1;
+
+    Vector3 gravity_g   = gravitational_acceleration(r);
+    Vector3 gravity_b   = q.inverse() * gravity_g;
+    Vector3 rdot        = q * v_b;
+    auto G              = L * H;
+    Vector4 qdot_coeffs = 0.5 * G * w;
+    Quaternion qdot{qdot_coeffs(0), qdot_coeffs(1), qdot_coeffs(2), qdot_coeffs(3)};
+    Vector3 vdot = gravity_b + f_b / mass - w.cross(v_b);
+    Vector3 wdot = Jinv * (tau_b - w.cross(J * w));
+
+    // the functions are called set_X but using them to set derivatives of X quantities in the correct positions of the
+    // state vector
+    ::set_pos_b_wrt_g_in_g(xdot, rdot);
+    ::set_g_q_b(xdot, qdot, false);
+    ::set_vel_b_wrt_g_in_b(xdot, vdot);
+    ::set_omega_b_wrt_g_in_b(xdot, wdot);
+
+    return xdot;
+}
+
+StateVector rk4(const StateVector& x, const Vector6& u, const Matrix_3x3& InertiaTensor,
```

Comment:
I prefer we use the RK4 defined in math library than this one. Reduces clutter

Reply:
I agree it would be nice to have standardized modules for stuff like this. I hardcoded the sizes of the vectors in this one to avoid dealing with dynamically sized matrices (I thought that would be slow). Admittedly, I have yet to test the impact on speed. Regardless, I don't foresee us doing more numerical integration, so there isn't a huge need for a general-purpose rk4.

Also, Amaar wanted pure dynamics and rk4 functions. Since the dynamics involve satellite parameters (mass, inertia, etc.), we need some way to get those into the state dynamics function, `f().` My approach is to add those mass/inertia variables to the parameter list. Hence, this rk4 is specialized and not general-purpose.